packages/editor/src/editor.ts

```diff
@@ -0,0 +1,236 @@
+        [Symbol.for('@@mongosh.syntheticPromise')]: true
+      });
+    };
+
+    wrapperFn.isDirectShellCommand = true;
+    wrapperFn.returnsPromise = true;
+    (instanceState.shellApi as any).edit = instanceState.context.edit = wrapperFn;
+    (signatures.ShellApi.attributes as any).edit = {
+      type: 'function',
+      returnsPromise: true,
+      isDirectShellCommand: true
+    } as TypeSignature;
+  }
+
+  activate(): void {
+    this._instanceState.registerPlugin(this);
+  }
+
+  // In the case of using VSCode as an external editor,
+  // we should detect whether the MongoDB extension is installed and open a .mongodb file.
+  // In all other cases we should open a .js file.
+  async _getExtension(cmd: string): Promise<string> {
+    if (!this._isVscodeApp(cmd)) {
+      return 'js';
+    }
+
+    try {
+      const extensions = await fse.readdir(path.join(this._vscodeDir, 'extensions'));
+      const hasMongodbExtension = !!extensions
+        .find((name) => name.includes('mongodb.mongodb-vscode'));
+
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-done', {
+        vscodeDir: this._vscodeDir,
+        hasMongodbExtension
+      });
+
+      return hasMongodbExtension ? 'mongodb' : 'js';
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-failed', {
+        action: 'mongosh-editor:read-vscode-extensions-failed',
+        error: error.message
+      });
+
+      return 'js';
+    }
+  }
+
+  async _getEditor(): Promise<string|null> {
+    // Check for an external editor in the mongosh configuration.
+    let editor: string | null = await this._instanceState.shellApi.config.get('editor');
+
+    // Check for an external editor in the environment variable.
+    if (!editor && process.env.EDITOR) {
+      editor = process.env.EDITOR;
+    }
+
+    return editor;
+  }
+
+  async _createTempFile(): Promise<void> {
+    const ext = await this._getExtension(this._editorSession.cmd);
+    const content = await this._getEditorContent();
+
+    this._editorSession.tmpDoc = path.join(this._tmpDir, `edit-${uuidv4()}.${ext}`);
+
+    // Create a temp file to store content that is being edited.
+    await fse.ensureFile(this._editorSession.tmpDoc);
+    await fse.writeFile(this._editorSession.tmpDoc, beautify(content));
+  }
+
+  async _readTempFile(tmpDoc: string): Promise<string> {
+    const content = await fse.readFile(tmpDoc, 'utf8');
+    await fse.unlink(tmpDoc);
+    return this._makeMultilineJSIntoSingleLine(content);
+  }
+
+  _isVscodeApp(cmd: string): boolean {
+    const regex = /^((.*[\/\\])*)?[cC]ode(.exe)?$/;
+
+    try {
+      return !!regex.exec(cmd);
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:is-vscode-app-regex-failed', { regex, cmd });
+      return false;
+    }
+  }
+
+  _isStatement(code: string): boolean {
+    const regex = /\b[^()]+\((.*)\)$/;
+
+    try {
+      return !!regex.exec(code);
+    } catch (error) {
```

Comment:
I might be wrong, but I don’t think `regex.exec()` can throw (and `regex.test` might be a bit nicer because that already returns a boolean :))

Reply:
I think it can, but I will replace it with `test` because you are right it is already a boolean then :)

Reply:
There’s nothing in the spec that I could see throwing inside `.exec()`, unless you call it on a non-RegExp `this` value or with an argument that cannot be converted to string (e.g. `{ toString() { throw new Error }}`) … also, if `.exec()` could throw, tons of code that use it would be broken :)